gnucash/import-export/qif-imp/qif-to-gnc.scm

```diff
@@ -449,6 +455,87 @@
           (qif-file:xtns qif-file)))
        sorted-qif-files-list)
 
+      ;; Add prices
+      (let ((pricedb (gnc-pricedb-get-db (gnc-get-current-book)))
+            (symbol-hash (make-symbol-hash stock-map))
+            (sorted-prices '())
+            (pruned-prices '())
+            (pruned-price-count 0)
+            (commodity #f)
+            (current-symbol " invalid symbol "))
+
+        ;; Let's combine all the prices and sort them; it
+        ;;  is slightly more efficient to do them by symbol and
+        ;;  tidies up error reporting on missing symbols
+        (for-each
+          (lambda (qif-file)
+            (set! sorted-prices (append sorted-prices (qif-file:prices qif-file))))
+         qif-files-list)
+
+        (set! sorted-prices (sort sorted-prices
+          (lambda (a b)
+            (string-ci<? (qif-price:symbol a) (qif-price:symbol b)))))
+
+        (gnc-pricedb-begin-edit pricedb)
+        ;; Turning off bulk update avoids an n^2 performance
+        ;;  cost, but also turns off duplicate checking
+        (gnc-pricedb-set-bulk-update pricedb #t)
+
+        (if progress-dialog
+            (gnc-progress-dialog-set-sub progress-dialog
+                (string-append (G_ "Discarding duplicate prices"))))
+
+        ;; Prune duplicate records before inserting
+        (for-each
+          (lambda (price)
+            (if (not (string-ci=? current-symbol (qif-price:symbol price)))
+                (begin
+                  (set! current-symbol (qif-price:symbol price))
+                  (set! commodity (hash-ref symbol-hash current-symbol))
+
+                  (if (equal? commodity #f)
+                      (qif-import:log progress-dialog
+                          "qif-import:qif-to-gnc"
+                          (G_ (format #f "Warning: cannot find commodity for symbol ~a." current-symbol))))))
+            (if (not (equal? commodity #f))
+                (if (not (duplicate-price? pricedb commodity default-currency price))
+                    (set! pruned-prices (append pruned-prices (list price)))
+                    (set! pruned-price-count (+ 1 pruned-price-count))))
+            (update-progress))
+         sorted-prices)
```

Comment:
```suggestion

        ;; Prune duplicate records before inserting
        (let lp ((sorted-prices sorted-prices)
                 (pruned '())
                 (count 0))
          (cond
           ((null? sorted-prices)
            (set! pruned-prices (reverse pruned))
            (set! pruned-price-count count))
           (else
            (update-progress)
            (unless (string-ci=? current-symbol (qif-price:symbol price))
              (set! current-symbol (qif-price:symbol price))
              (set! commodity (hash-ref symbol-hash current-symbol))
              (unless commodity
                (qif-import:log progress-dialog
                                "qif-import:qif-to-gnc"
                                (G_ (format #f "Warning: cannot find commodity for symbol ~a." current-symbol)))))
            (when commodity
              (if (duplicate-price? pricedb commodity default-currency price)
                  (lp (cdr sorted-prices)
                      pruned
                      (1+ count))
                  (lp (cdr sorted-prices)
                      (cons price pruned)
                      count))))))
```

Reply:
I'm confused.  I don't see an O(n^2) problem.  I have 70k records that are pruned in about 29 seconds, which feels reasonable to me.  The suggestion as provided does not work.  It does not do a (let ((price (cdr sorted-prices)) ..., and the first unless needs a call to (lp (cdr..., to keep the logic going.  And then the logic flow does not work when there are multiple prices with a bad symbol in a row (in fact, it incorrectly exits early), whereas my code pruned those just fine.  Can you explain the flaw you saw so that I can try to address it?  Is (append ...) expensive in some way I don't understand?

Reply:
https://chat.openai.com/share/4f5758a1-1e00-4864-8e71-8f3903d7eedf ;)

Reply:
If you wish to retain your code structure, then simply build pruned-list via `(set! pruned-prices (cons price pruned-prices))` then when the forloop completes, `(set! pruned-prices (reverse pruned-prices))`

Reply:
@jwhite66 do you understand *why* it's faster?

Reply:
Well, I presume it's because the internal implementation of a scheme list is a singly linked list, and so append requires a full list traversal on each call.  For the record, I think the other point Christopher was making was that my code is not idiomatic scheme.  That is unsurprising, as it's my first substantive bit of scheme :-/.   I would prefer the code go in as I wrote it because I believe I understand it, and have some comfort that I have developed and tested those code paths as written, no matter how naive.   I did study the code shared and felt that I should embrace at least when and unless.  I never really enjoyed the recursion first approach of lisp like languages, so that may not come so readily to me.

Reply:
Yes, that's right. 

I understand your preference but we need to be concerned about maintaining your contribution going forward, meaning that other people--maybe even your future self who's learned more scheme but forgotten this PR--can easily understand it and how it fits in with the rest of the importer.

Reply:
Sure, and as maintainers you have a duty to make sure the code base has the quality it needs, and I am willing to make changes until it passes muster.  From my naive and definitely inexperienced perspective, it seems to be Good Enough (TM).  The qif import code itself feels a bit on the naive side, so my code seemed to fit right in :-/.  Again, I'm willing to make changes, but I would need more specific direction on which parts were obviously wrong.